glommio/src/executor/stall.rs

```diff
@@ -0,0 +1,534 @@
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2022 Datadog, Inc.
+//
+
+use crate::executor::TaskQueueHandle;
+use nix::sys;
+use std::{
+    fmt,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    thread::{self, JoinHandle},
+    time::{Duration, Instant},
+};
+
+pub struct StallDetection<'a> {
+    executor: usize,
+    queue_handle: TaskQueueHandle,
+    queue_name: &'a str,
+    trace: backtrace::Backtrace,
+    budget: Duration,
+    overage: Duration,
+}
+
+impl fmt::Debug for StallDetection<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("StallDetection")
+            .field("executor", &self.executor)
+            .field("queue_handle", &self.queue_handle)
+            .field("queue_name", &self.queue_name)
+            .field("trace", &self.trace)
+            .field("budget", &self.budget)
+            .field("overage", &self.overage)
+            .finish()
+    }
+}
+
+impl fmt::Display for StallDetection<'_> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "[stall-detector -- executor {}] task queue {} went over-budget: {:#?} (budget: \
+             {:#?}). Backtrace: {:#?}",
+            self.executor, self.queue_name, self.overage, self.budget, self.trace,
+        )
+    }
+}
+
+/// Trait describing what signal to use to trigger stall detection,
+/// how far past expected execution time to trigger a stall,
+/// and how to handle a stall detection once triggered.
+pub trait StallDetectionHandler: std::fmt::Debug + Send + Sync {
+    /// How far past the preemption timer should qualify as a stall
+    /// If None is returned, don't use the stall detector for this task queue.
+    fn high_water_mark(
+        &self,
+        queue_handle: TaskQueueHandle,
+        max_expected_runtime: Duration,
+    ) -> Option<Duration>;
+
+    /// What signal number to use; see values in libc::SIG*
+    fn signal(&self) -> u8;
+
+    /// Handler called when a task exceeds its budget
+    ///
+    /// **NOTE**: This callback is invoked from inside the reactor.
+    /// Do not perform blocking syscalls, acquire locks, spawn tasks
+    /// or otherwise block here.
+    fn stall(&self, detection: StallDetection<'_>);
+}
+
+/// Default settings for signal number, high water mark and stall handler.
+/// By default, the high water mark to consider a task queue stalled is set to
+/// 10% of the expected run time. The default handler will log a stack trace of
+/// the currently executing task queue. The default signal number is
+/// [`nix::libc::SIGUSR1`].
+#[derive(Debug)]
+pub struct LoggingStallDetectionHandler {
+    /// What signal to use to trigger stall detection
+    pub signal: u8,
+    /// What percentage of the original expected runtime to allow before
+    /// triggering a stall
+    pub overage_ratio: f64,
```

Comment:
I am still not a fan of the float ratio.
My reason for that is that it is hard to reason about.

I see where you're coming from: because there can't be different latency goals, you'd likely be okay with a 30ms overshoot on a 100ms latency goal and only okay with a 3ms overshoot on a 10ms goal.

But there are two big problems with that:
1) It is essentially impossible to predict what the goal will be. The latency threshold is not a property of a task queue: it is a property of a *set of runnable task queues*. If there is a (runnable) task queue with a 100ms threshold and another with 1ms, then the threshold is 1ms. This comes and goes and I think it complicates the design.

2) In an ideal world, nobody ever goes over the task quota, at any time, for no period of time. (or an absolute threshold == 0) Of course such world doesn't exist, so we log those events. But it doesn't sound okay to me to overshoot the quota for 30ms just because your task queue has a higher threshold. Ultimately, the stall detector is not about your task queue, but about the fact that you're preventing other task queues to run by clogging the pipes.